Packages: Use esm for esm builds

sirreal opened the pull request:
esm build of packages were not using esm! Fix that.

Discovered in #36186 

<details>

<summary>Comparison of output</summary>

This is the result of transpiling `@automattic/calypso-ui` `ScreenReaderText` component.

```sh
npm run clean:packages && npm run build-packages
```

### CommonJS

#### Before

```js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ScreenReaderText;

var _react = _interopRequireDefault(require("react"));

require("./style.scss");

/**
 * External dependencies
 */

/**
 * Style dependencies
 */
function ScreenReaderText(_ref) {
  var children = _ref.children;
  return _react.default.createElement("span", {
    className: "screen-reader-text"
  }, children);
}
```

#### After

```js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ScreenReaderText;

var _react = _interopRequireDefault(require("react"));

require("./style.scss");

/**
 * External dependencies
 */

/**
 * Style dependencies
 */
function ScreenReaderText(_ref) {
  var children = _ref.children;
  return _react.default.createElement("span", {
    className: "screen-reader-text"
  }, children);
}
```

### ECMAScript Modules

#### Before

**It's not ESM!**

```js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ScreenReaderText;

var _react = _interopRequireDefault(require("react"));

require("./style.scss");

/**
 * External dependencies
 */

/**
 * Style dependencies
 */
function ScreenReaderText(_ref) {
  var children = _ref.children;
  return _react.default.createElement("span", {
    className: "screen-reader-text"
  }, children);
}
```

#### After

```js
/**
 * External dependencies
 */
import React from 'react';
/**
 * Style dependencies
 */

import './style.scss';
export default function ScreenReaderText(_ref) {
  var children = _ref.children;
  return React.createElement("span", {
    className: "screen-reader-text"
  }, children);
}
```

</details>

## Testing
1. Build packages: `npm run clean:packages && npm run build-packages`
1. Check the output of packages be inspecting the output in the `dist` directory.
1. Smoke test calypso.

sirreal committed:
Use esm for esm builds

sirreal committed:
Transpile esm correctly

sirreal commented:
Another comparison of `tree-select`:

<details>

<summary>CJS</summary>

#### Before

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = treeSelect;

/** @format */
var defaultGetCacheKey = function defaultGetCacheKey() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args.join();
};

var isFunction = function isFunction(fn) {
  return fn && typeof fn === 'function';
};

var isObject = function isObject(o) {
  return o && typeof o === 'object';
};
/**
 * Returns a selector that caches values.
 *
 * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
 *                                    Must return an array which gets passed as the first arg to the selector
 * @param  {Function} selector      A standard selector for calculating cached result
 * @param  {Object}   options       Options bag with additional arguments
 * @param  {Function} options.getCacheKey
 *                                  Custom way to compute the cache key from the `args` list
 * @return {Function}               Cached selector
 */


function treeSelect(getDependents, selector) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  if (process.env.NODE_ENV !== 'production') {
    if (!isFunction(getDependents) || !isFunction(selector)) {
      throw new TypeError('treeSelect: invalid arguments passed, selector and getDependents must both be functions');
    }
  }

  var cache = new WeakMap();
  var _options$getCacheKey = options.getCacheKey,
      getCacheKey = _options$getCacheKey === void 0 ? defaultGetCacheKey : _options$getCacheKey;

  var cachedSelector = function cachedSelector(state) {
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    var dependents = getDependents.apply(void 0, [state].concat(args));

    if (process.env.NODE_ENV !== 'production') {
      if (getCacheKey === defaultGetCacheKey && args.some(isObject)) {
        throw new Error('Do not pass objects as arguments to a treeSelector');
      }
    } // create a dependency tree for caching selector results.
    // this is beneficial over standard memoization techniques so that we can
    // garbage collect any values that are based on outdated dependents


    var leafCache = dependents.reduce(insertDependentKey, cache);
    var key = getCacheKey.apply(void 0, args);

    if (leafCache.has(key)) {
      return leafCache.get(key);
    }

    var value = selector.apply(void 0, [dependents].concat(args));
    leafCache.set(key, value);
    return value;
  };

  cachedSelector.clearCache = function () {
    // WeakMap doesn't have `clear` method, so we need to recreate it
    cache = new WeakMap();
  };

  return cachedSelector;
}
/*
 * This object will be used as a WeakMap key if a dependency is a falsy value (null, undefined, ...)
 */


var STATIC_FALSY_KEY = {};
/*
 * First tries to get the value for the key.
 * If the key is not present, then inserts a new map and returns it
 *
 * Note: Inserts WeakMaps except for the last map which will be a regular Map.
 * The last map is a regular one because the the key for the last map is the string results of args.join().
 */

function insertDependentKey(map, key, currentIndex, arr) {
  var weakMapKey = key || STATIC_FALSY_KEY;
  var existingMap = map.get(weakMapKey);

  if (existingMap) {
    return existingMap;
  }

  var newMap = currentIndex === arr.length - 1 ? new Map() : new WeakMap();
  map.set(weakMapKey, newMap);
  return newMap;
}
```

#### After

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = treeSelect;

/** @format */
var defaultGetCacheKey = function defaultGetCacheKey() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args.join();
};

var isFunction = function isFunction(fn) {
  return fn && typeof fn === 'function';
};

var isObject = function isObject(o) {
  return o && typeof o === 'object';
};
/**
 * Returns a selector that caches values.
 *
 * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
 *                                    Must return an array which gets passed as the first arg to the selector
 * @param  {Function} selector      A standard selector for calculating cached result
 * @param  {Object}   options       Options bag with additional arguments
 * @param  {Function} options.getCacheKey
 *                                  Custom way to compute the cache key from the `args` list
 * @return {Function}               Cached selector
 */


function treeSelect(getDependents, selector) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  if (process.env.NODE_ENV !== 'production') {
    if (!isFunction(getDependents) || !isFunction(selector)) {
      throw new TypeError('treeSelect: invalid arguments passed, selector and getDependents must both be functions');
    }
  }

  var cache = new WeakMap();
  var _options$getCacheKey = options.getCacheKey,
      getCacheKey = _options$getCacheKey === void 0 ? defaultGetCacheKey : _options$getCacheKey;

  var cachedSelector = function cachedSelector(state) {
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    var dependents = getDependents.apply(void 0, [state].concat(args));

    if (process.env.NODE_ENV !== 'production') {
      if (getCacheKey === defaultGetCacheKey && args.some(isObject)) {
        throw new Error('Do not pass objects as arguments to a treeSelector');
      }
    } // create a dependency tree for caching selector results.
    // this is beneficial over standard memoization techniques so that we can
    // garbage collect any values that are based on outdated dependents


    var leafCache = dependents.reduce(insertDependentKey, cache);
    var key = getCacheKey.apply(void 0, args);

    if (leafCache.has(key)) {
      return leafCache.get(key);
    }

    var value = selector.apply(void 0, [dependents].concat(args));
    leafCache.set(key, value);
    return value;
  };

  cachedSelector.clearCache = function () {
    // WeakMap doesn't have `clear` method, so we need to recreate it
    cache = new WeakMap();
  };

  return cachedSelector;
}
/*
 * This object will be used as a WeakMap key if a dependency is a falsy value (null, undefined, ...)
 */


var STATIC_FALSY_KEY = {};
/*
 * First tries to get the value for the key.
 * If the key is not present, then inserts a new map and returns it
 *
 * Note: Inserts WeakMaps except for the last map which will be a regular Map.
 * The last map is a regular one because the the key for the last map is the string results of args.join().
 */

function insertDependentKey(map, key, currentIndex, arr) {
  var weakMapKey = key || STATIC_FALSY_KEY;
  var existingMap = map.get(weakMapKey);

  if (existingMap) {
    return existingMap;
  }

  var newMap = currentIndex === arr.length - 1 ? new Map() : new WeakMap();
  map.set(weakMapKey, newMap);
  return newMap;
}
```

</details>

<details>

<summary>ESM</summary>

#### Before

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = treeSelect;

/** @format */
var defaultGetCacheKey = function defaultGetCacheKey() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args.join();
};

var isFunction = function isFunction(fn) {
  return fn && typeof fn === 'function';
};

var isObject = function isObject(o) {
  return o && typeof o === 'object';
};
/**
 * Returns a selector that caches values.
 *
 * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
 *                                    Must return an array which gets passed as the first arg to the selector
 * @param  {Function} selector      A standard selector for calculating cached result
 * @param  {Object}   options       Options bag with additional arguments
 * @param  {Function} options.getCacheKey
 *                                  Custom way to compute the cache key from the `args` list
 * @return {Function}               Cached selector
 */


function treeSelect(getDependents, selector) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  if (process.env.NODE_ENV !== 'production') {
    if (!isFunction(getDependents) || !isFunction(selector)) {
      throw new TypeError('treeSelect: invalid arguments passed, selector and getDependents must both be functions');
    }
  }

  var cache = new WeakMap();
  var _options$getCacheKey = options.getCacheKey,
      getCacheKey = _options$getCacheKey === void 0 ? defaultGetCacheKey : _options$getCacheKey;

  var cachedSelector = function cachedSelector(state) {
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    var dependents = getDependents.apply(void 0, [state].concat(args));

    if (process.env.NODE_ENV !== 'production') {
      if (getCacheKey === defaultGetCacheKey && args.some(isObject)) {
        throw new Error('Do not pass objects as arguments to a treeSelector');
      }
    } // create a dependency tree for caching selector results.
    // this is beneficial over standard memoization techniques so that we can
    // garbage collect any values that are based on outdated dependents


    var leafCache = dependents.reduce(insertDependentKey, cache);
    var key = getCacheKey.apply(void 0, args);

    if (leafCache.has(key)) {
      return leafCache.get(key);
    }

    var value = selector.apply(void 0, [dependents].concat(args));
    leafCache.set(key, value);
    return value;
  };

  cachedSelector.clearCache = function () {
    // WeakMap doesn't have `clear` method, so we need to recreate it
    cache = new WeakMap();
  };

  return cachedSelector;
}
/*
 * This object will be used as a WeakMap key if a dependency is a falsy value (null, undefined, ...)
 */


var STATIC_FALSY_KEY = {};
/*
 * First tries to get the value for the key.
 * If the key is not present, then inserts a new map and returns it
 *
 * Note: Inserts WeakMaps except for the last map which will be a regular Map.
 * The last map is a regular one because the the key for the last map is the string results of args.join().
 */

function insertDependentKey(map, key, currentIndex, arr) {
  var weakMapKey = key || STATIC_FALSY_KEY;
  var existingMap = map.get(weakMapKey);

  if (existingMap) {
    return existingMap;
  }

  var newMap = currentIndex === arr.length - 1 ? new Map() : new WeakMap();
  map.set(weakMapKey, newMap);
  return newMap;
}
```

#### After


```js
/** @format */
var defaultGetCacheKey = function defaultGetCacheKey() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args.join();
};

var isFunction = function isFunction(fn) {
  return fn && typeof fn === 'function';
};

var isObject = function isObject(o) {
  return o && typeof o === 'object';
};
/**
 * Returns a selector that caches values.
 *
 * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
 *                                    Must return an array which gets passed as the first arg to the selector
 * @param  {Function} selector      A standard selector for calculating cached result
 * @param  {Object}   options       Options bag with additional arguments
 * @param  {Function} options.getCacheKey
 *                                  Custom way to compute the cache key from the `args` list
 * @return {Function}               Cached selector
 */


export default function treeSelect(getDependents, selector) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  if (process.env.NODE_ENV !== 'production') {
    if (!isFunction(getDependents) || !isFunction(selector)) {
      throw new TypeError('treeSelect: invalid arguments passed, selector and getDependents must both be functions');
    }
  }

  var cache = new WeakMap();
  var _options$getCacheKey = options.getCacheKey,
      getCacheKey = _options$getCacheKey === void 0 ? defaultGetCacheKey : _options$getCacheKey;

  var cachedSelector = function cachedSelector(state) {
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }

    var dependents = getDependents.apply(void 0, [state].concat(args));

    if (process.env.NODE_ENV !== 'production') {
      if (getCacheKey === defaultGetCacheKey && args.some(isObject)) {
        throw new Error('Do not pass objects as arguments to a treeSelector');
      }
    } // create a dependency tree for caching selector results.
    // this is beneficial over standard memoization techniques so that we can
    // garbage collect any values that are based on outdated dependents


    var leafCache = dependents.reduce(insertDependentKey, cache);
    var key = getCacheKey.apply(void 0, args);

    if (leafCache.has(key)) {
      return leafCache.get(key);
    }

    var value = selector.apply(void 0, [dependents].concat(args));
    leafCache.set(key, value);
    return value;
  };

  cachedSelector.clearCache = function () {
    // WeakMap doesn't have `clear` method, so we need to recreate it
    cache = new WeakMap();
  };

  return cachedSelector;
}
/*
 * This object will be used as a WeakMap key if a dependency is a falsy value (null, undefined, ...)
 */

var STATIC_FALSY_KEY = {};
/*
 * First tries to get the value for the key.
 * If the key is not present, then inserts a new map and returns it
 *
 * Note: Inserts WeakMaps except for the last map which will be a regular Map.
 * The last map is a regular one because the the key for the last map is the string results of args.join().
 */

function insertDependentKey(map, key, currentIndex, arr) {
  var weakMapKey = key || STATIC_FALSY_KEY;
  var existingMap = map.get(weakMapKey);

  if (existingMap) {
    return existingMap;
  }

  var newMap = currentIndex === arr.length - 1 ? new Map() : new WeakMap();
  map.set(weakMapKey, newMap);
  return newMap;
}
```

</details>

sirreal commented:
I'm investigating the reason for CI failures. Local builds seem to work fine.

sirreal commented:
I'll come back to this later. I'm unable to reproduce locally including in docker.

sirreal commented on packages/calypso-build/babel/default.js:
```diff
@@ -21,7 +28,7 @@ module.exports = () => ( {
 				helpers: true,
 				regenerator: false,
-				useESModules: false,
+				useESModules: modules === false,
```
I ended up dropping this change because it seemed to break things.

jsnajdr commented:
> I'll come back to this later. I'm unable to reproduce locally including in docker.

I can reproduce the errors locally. Investigating.

jsnajdr committed:
Remove duplicate usage of Babel Env preset

The `calypso-build` default preset does `require.resolve( '@babel/preset-env` )`, and then
out main config loads the same preset again, but by specifying it as `@babel/preset-env` string.
That defeats the internal Babel deduplication and we end up with the preset applied twice.
That leads to very weird build errors in the production environment.

Add a `modules` option to the `calypso-build` preset and pass it from Calypso Babel config instead
of overriding with a new instance of the Env preset.

Also, it makes sense for `modules: false` to be the default. For Babel, that means that it should
not transpile `import` and `export` statement and should leave them as-is. Only when specifying
something non-default, e.g., `modules: 'commonjs'`, we tell Babel to transpile things. Changed
the code accordingly.

jsnajdr commented:
The production build:
```
CALYPSO_ENV=production NODE_ENV=production npm run build
```
was failing with really weird errors with this PR applied. What happened?

The `calypso-build` default preset does `require.resolve( '@babel/preset-env` )`, and then our main config loads the same preset again, but by specifying it as `@babel/preset-env` string. That defeats the internal Babel deduplication and we end up with the preset applied twice. That's very unexpected and weird things start happening.

The simplest fix is to use `require.resolve( '@babel/preset-env` )` in the top-level Calypso Babel config, too, and the build starts working.

A better fix is to add a `modules` option to the `calypso-build` preset and pass it from Calypso Babel config instead of overriding with a new instance of the Env preset. Did that in fc9e6eb.

Also, it makes sense for `modules: false` to be the default. For Babel, that means that it should not transpile `import` and `export` statement and should leave them as-is. Only when specifying something non-default, e.g., `modules: 'commonjs'`, we tell Babel to transpile things. Changed
the code accordingly.

Let's see how it goes.

jsnajdr reviewed (approved):
Things are green now and I think this is ready to :ship:

It would be nice for someone to cross-check my changes so that I'm not approving my own code.

sirreal commented:
Nice detective work @jsnajdr 👏 The changes make sense to me, consider this my approval if you'd like to ship. We'll need to bump package versions and publish and update to each of them with correct `esm` builds.

I wonder if previous builds were preventing me from detecting the problems locally or in Docker 🤔

jsnajdr committed:
Add documentation to README and a changelog entry

jsnajdr commented:
> We'll need to bump package versions and publish and update to each of them

I'd also like us to stop shipping the `test` subdirectories as part of the package.

sirreal commented:
> I'd also like us to stop shipping the `test` subdirectories as part of the package.

I agree, but that can be tackled independently. I think we should considering using `tsc` for all of our package builds. I've been working on that in https://github.com/Automattic/wp-calypso/pull/36186

jsnajdr commented:
> I agree, but that can be tackled independently.

@sirreal I just did it in #36289, without waiting for `tsc`. Please :eyes: when you have a moment.

ljharb commented:
Heads up that my tests on #36256 are failing after rebasing past this change, with an `import` error in jest.

sirreal commented:
> Heads up that my tests on #36256 are failing after rebasing past this change, with an `import` error in jest.

Thanks for the report! Proposed fix: https://github.com/Automattic/wp-calypso/pull/36349

ockham commented:
I filed a PR to update Jetpack to `calypso-build` 5.0.0, https://github.com/Automattic/jetpack/pull/13854. Turns out this needed quite a few changes because of the defaulting to ESM that this PR introduced; the problem is mostly with individual tools' config files that are read by node, and thus need their `import`s transpiled to `require`s.

I'm a bit worried it might be an obstacle for `calypso-build` users, since some of the required changes are a bit subtle; and in some places, I found it rather unintuitive _where_ I had to add the `modules: 'commonjs' options. For example, I had to add it to `tests/jest-extensions-babel-transform.js`, even though we're using `calypso-build`'s Jest preset, which is [supposed to do just that](https://github.com/Automattic/wp-calypso/blob/fba2d80dc733828ff3fff47af86f0944cca83bff/packages/calypso-build/jest/transform/babel.js#L9) ([source](https://github.com/Automattic/jetpack/pull/13854#issuecomment-547093346)).

I wonder if there's anything we can do in `calypso-build` to require fewer changes from its users when they upgrade to `calypso-build` >= 4.2.0 :thinking:

sirreal commented:
The provided jest preset works for the "default" build system (no `--env.WP`).

They WordPress build setup includes the babel transform that adds `import { createElement } from '@wordpress/element'` as required by JSX. That works well with our webpack wrapper that allows arbitrary arguments.

Jest has no such configuration, its arguments are close and I believe it warns or errors on unknown arguments. What's required is some way to use a set of babel transforms in our jest transform that includes the transform I mentioned above.

Possibilities I see:
- This could be a single preset that reads an environment variable to include or exclude the WordPress transforms (`JEST_ENV_WP=true jest`). Putting more logic into arbitrary environment variables isn't very appealing.
- Providing a different preset. The default would remain available:
  ```json
  { "preset": "@automattic/calypso-build" }
  ```
  and we'd add a preset for WP type builds:
  ```json
  { "preset": "@automattic/calypso-build/jest-preset-wp" }
  ```

I have a hard time imagining other alternatives without creating a jest wrapper, and that doesn't seem like an appealing alternative.
